File: src/commands/plots/timelaps.js
const { SlashCommandBuilder, AttachmentBuilder } = require('discord.js');
const { spawn } = require('child_process');
const { downloadData } = require('../../utils/functions');

module.exports = {
	data: new SlashCommandBuilder()
		.setName('timelaps')
		.setDescription('Create a graph of the laps')
		.addStringOption(option =>
			option.setName('year')
				.setDescription('The year of the GP')
				.setRequired(true))
		.addStringOption(option =>
			option.setName('gp')
				.setDescription('The GP you want to analyse')
				.setRequired(true))
		.addStringOption(option =>
			option.setName('session')
				.setDescription('The session you want to analyse (FP1,FP2,FP3 for Practice; Q for Qualification; R for Race)')
				.setRequired(true))
		.addStringOption(option =>
			option.setName('driver1')
				.setDescription('The TLA of the 1st driver you want to analyse')
				.setRequired(true))
		.addStringOption(option =>
			option.setName('driver2')
				.setDescription('The TLA of the 2nd driver you want to analyse')
				.setRequired(true))
		.addStringOption(option =>
			option.setName('color_driver1')
				.setDescription('The Color for the 1st driver you want to analyse (b,r,g,c,m,y,k,w)')
				.setRequired(false))
		.addStringOption(option =>
			option.setName('color_driver2')
				.setDescription('The Color for the 2nd driver you want to analyse (b,r,g,c,m,y,k,w)')
				.setRequired(false)),

	async execute(interacton, client) {
		downloadData(interacton);
	},

	async python(interaction) {
		let Year = interaction.options.getString('year');
		let GP = interaction.options.getString('gp');
		let Session = interaction.options.getString('session').toUpperCase();
		let Driver1 = interaction.options.getString('driver1').toUpperCase();
		let Driver2 = interaction.options.getString('driver2').toUpperCase();
		let ColorDriver1 = interaction.options.getString('color_driver1');
		let ColorDriver2 = interaction.options.getString('color_driver2');
		
		if(ColorDriver1 === null) ColorDriver1 = 'default';
		if(ColorDriver2 === null) ColorDriver2 = 'default';
		
		const python = spawn('python3', ['src/commands/plots/python/timelaps.py', Year, GP, Session, Driver1, Driver2, ColorDriver1, ColorDriver2]);
		
		let errors = [];

		python.stdout.on('data', (data) => {
			console.log(`stdout: ${data}`);
		});
		python.stderr.on('data', (data) => {
			console.error(`stderr: ${data}`);
			errors.push(data);
		});

		python.on('close', async (code) => {
			console.log(`child process close all stdio with code ${code}`);
			if(code !== 0) {
				let _error = errors[errors.length-1];
				await interaction.editReply(`Sorry, this command failed. ${_error.includes('The data you are trying to access has not been loaded yet. See `Session.load`') ? 'The data is not available for now.' : `Python ERROR \`\`\`${_error}\`\`\``}. `);
				return;
			}
			const attachment = new AttachmentBuilder('src/commands/plots/python/Figure.png', { name: 'Plot.png' });
			await interaction.editReply({content: `Lap times comparison ${Driver1} vs. ${Driver2} - ${GP} ${Year} ${Session}`, files: [attachment]});
		});
	}
};

File: src/commands/plots/python/timelaps.py
import fastf1 as ff1
from fastf1 import plotting
from matplotlib import pyplot as plt
import sys

from requests import session

plotting.setup_mpl()

# Enable the cache
ff1.Cache.enable_cache('src/commands/plots/python/cache') # Argument is the name of the folder

Year = int(sys.argv[1])
GP = sys.argv[2]
Session = sys.argv[3]

Driver1 = sys.argv[4]
Driver2 = sys.argv[5]

session = ff1.get_session(Year, GP, Session)
session.load()

Driver1_laps = session.laps.pick_driver(Driver1)
Driver2_laps = session.laps.pick_driver(Driver2)

team_driver_1 = Driver1_laps['Team'].iloc[0]
team_driver_2 = Driver2_laps['Team'].iloc[0]

try:
    color_driver_1 = sys.argv[6]
    if sys.argv[6] == "default":
        color_driver_1 = ff1.plotting.team_color(team_driver_1)
except:
    color_driver_1 = ff1.plotting.team_color(team_driver_1)

try:
    color_driver_2 = sys.argv[7]
    if sys.argv[7] == "default":
        color_driver_2 = ff1.plotting.team_color(team_driver_2)
except:
    color_driver_2 = ff1.plotting.team_color(team_driver_2)


fig, ax = plt.subplots()
ax.plot(Driver1_laps['LapNumber'], Driver1_laps['LapTime'], label=Driver1, color=color_driver_1) # type: ignore
ax.plot(Driver2_laps['LapNumber'], Driver2_laps['LapTime'], label=Driver2, color=color_driver_2) # type: ignore
ax.set_title(f"Lap times comparison {Driver1} vs. {Driver2} - {GP} {Year} {Session}", {'fontsize': 11})
ax.set_xlabel("Lap Number\nF1 Telemetry Discord Bot - Developed by LapsTime, Flip and Walshaw", {'fontsize': 7.5})
ax.set_ylabel("Lap Time")
ax.legend(loc="upper left")
plt.savefig('src/commands/plots/python/Figure.png')